src/main/java/bio/terra/service/snapshotbuilder/utils/CriteriaQueryBuilder.java

```diff
@@ -0,0 +1,211 @@
+
+    TablePointer ancestorPointer =
+        TablePointer.fromTableName("concept_ancestor", tableNameGenerator);
+    TableVariable ancestorVariable =
+        TableVariable.forJoined(
+            ancestorPointer,
+            "ancestor_concept_id",
+            new FieldVariable(
+                new FieldPointer(occurrencePointer, occurrenceTable.idColumnName()),
+                occurrenceVariable));
+
+    return new SubQueryFilterVariable(
+        new FieldVariable(
+            new FieldPointer(getRootTablePointer(), PERSON_ID_FIELD_NAME), getRootTableVariable()),
+        SubQueryFilterVariable.Operator.IN,
+        new Query(
+            List.of(
+                new FieldVariable(
+                    new FieldPointer(occurrencePointer, PERSON_ID_FIELD_NAME), occurrenceVariable)),
+            List.of(occurrenceVariable, ancestorVariable),
+            new BooleanAndOrFilterVariable(
+                BooleanAndOrFilterVariable.LogicalOperator.OR,
+                List.of(
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(occurrencePointer, occurrenceTable.idColumnName()),
+                            occurrenceVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().intValue())),
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(ancestorPointer, "ancestor_concept_id"),
+                            ancestorVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().intValue()))))));
+  }
+
+  public FilterVariable generateFilterForCriteria(SnapshotBuilderCriteria criteria) {
+    return switch (criteria.getKind()) {
+      case LIST -> {
+        if (criteria instanceof SnapshotBuilderProgramDataListCriteria listCriteria) {
+          yield generateFilterForListCriteria(listCriteria);
+        }
+        throw new BadRequestException("Malformed list criteria");
+      }
+      case RANGE -> {
+        if (criteria instanceof SnapshotBuilderProgramDataRangeCriteria rangeCriteria) {
+          yield generateFilterForRangeCriteria(rangeCriteria);
+        }
+        throw new BadRequestException("Malformed range criteria");
+      }
+      case DOMAIN -> {
+        if (criteria instanceof SnapshotBuilderDomainCriteria domainCriteria) {
+          yield generateFilterForDomainCriteria(domainCriteria);
+        }
+        throw new BadRequestException("Malformed domain criteria");
+      }
+    };
+  }
+
+  public FilterVariable generateAndOrFilterForCriteriaGroup(
+      SnapshotBuilderCriteriaGroup criteriaGroup) {
+    return new BooleanAndOrFilterVariable(
+        criteriaGroup.isMeetAll() != null && criteriaGroup.isMeetAll()
```

Comment:
You can use `requireNonNull()` to combine the null check and its default behavior
```suggestion
        Objects.requireNonNullElse(criteriaGroup.isMeetAll(), false)
```

But I wonder why this null check is needed at all. I would expect these fields to be required and never null.